{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "87321964",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "721f4c41",
   "metadata": {},
   "outputs": [],
   "source": [
    "from mjol import gan as mgan"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "75ccc655",
   "metadata": {},
   "outputs": [],
   "source": [
    "fn = \"/ccb/salz4-3/hji20/hg002-q100-annotation/data/chm13v2.0_RefSeq_Liftoff_v5.2.gff3\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a141e4fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "gan = mgan.GAn(\n",
    "    filename = fn,\n",
    "    format = 'gff'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d739ca3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "exon\n",
      "cds\n"
     ]
    }
   ],
   "source": [
    "gan.build_db(use_iv = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "23ef628c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>src</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>gene</td>\n",
       "      <td>6047</td>\n",
       "      <td>13941</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=LOC124900618;gene_name=LOC124900618;db_xref...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>transcript</td>\n",
       "      <td>6047</td>\n",
       "      <td>13941</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=XR_002958507.2;Parent=LOC124900618;db_xref=...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>exon</td>\n",
       "      <td>6047</td>\n",
       "      <td>6420</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=XR_002958507.2;db_xref=GeneID:124900618...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>exon</td>\n",
       "      <td>12078</td>\n",
       "      <td>12982</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=XR_002958507.2;db_xref=GeneID:124900618...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>exon</td>\n",
       "      <td>13445</td>\n",
       "      <td>13579</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=XR_002958507.2;db_xref=GeneID:124900618...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr      src feature_type  start    end score strand frame  \\\n",
       "0  chr1  Liftoff         gene   6047  13941     .      -     .   \n",
       "1  chr1  Liftoff   transcript   6047  13941     .      -     .   \n",
       "2  chr1  Liftoff         exon   6047   6420     .      -     .   \n",
       "3  chr1  Liftoff         exon  12078  12982     .      -     .   \n",
       "4  chr1  Liftoff         exon  13445  13579     .      -     .   \n",
       "\n",
       "                                          attributes  \n",
       "0  ID=LOC124900618;gene_name=LOC124900618;db_xref...  \n",
       "1  ID=XR_002958507.2;Parent=LOC124900618;db_xref=...  \n",
       "2  Parent=XR_002958507.2;db_xref=GeneID:124900618...  \n",
       "3  Parent=XR_002958507.2;db_xref=GeneID:124900618...  \n",
       "4  Parent=XR_002958507.2;db_xref=GeneID:124900618...  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df = pd.read_csv(fn, sep='\\t', comment='#', header=None)\n",
    "df.columns = mgan.HDR\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0bfdf3da",
   "metadata": {},
   "outputs": [],
   "source": [
    "from mjol import base as mbase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1cd083d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "tx = None\n",
    "for _, row in df.iterrows():\n",
    "    if row['feature_type'] == \"transcript\":\n",
    "        if tx is not None:\n",
    "            break\n",
    "        tx = mbase.IntervalGFeature(\n",
    "                chr = row['chr'],\n",
    "                src = row['src'],\n",
    "                feature_type = row['feature_type'],\n",
    "                start = row['start'],\n",
    "                end = row['end'],\n",
    "                score = None if row['score'] == '.' else row['score'],\n",
    "                strand = row['strand'],\n",
    "                frame = row['frame'],\n",
    "                attributes = mbase.load_attributes(row['attributes'], \n",
    "                                        kv_sep = '='),\n",
    "                children = []\n",
    "            )\n",
    "    elif row['feature_type'] == \"exon\":\n",
    "        exon = mbase.GFeature(\n",
    "                chr = row['chr'],\n",
    "                src = row['src'],\n",
    "                feature_type = row['feature_type'],\n",
    "                start = row['start'],\n",
    "                end = row['end'],\n",
    "                score = None if row['score'] == '.' else row['score'],\n",
    "                strand = row['strand'],\n",
    "                frame = row['frame'],\n",
    "                attributes = mbase.load_attributes(row['attributes'], \n",
    "                                        kv_sep = '='),\n",
    "                children = []\n",
    "            )\n",
    "        tx.add_a_child(exon)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "e66dd05a",
   "metadata": {},
   "outputs": [],
   "source": [
    "tx.build_itree_from_children()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c381a6b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{Interval(12078, 12982, exon:Supporting evidence includes similarity to: 1 long SRA read%2C and 90%25 coverage of the annotated genomic feature by RNAseq alignments%2C including 1 sample with support for all annotated introns,-,12078-12982)}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tx.query_itree(12100, 12200)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mjol",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
